### DISCUSSION TITLE
Add way to programmatically base value of one field on the value of another

### DISCUSSION THREAD

**Discussion by maartenvn** [directus/directus#4231] [Feature Requests]
**Directus**: 9.0.0-beta.2
**Node**: v14.10.0
**MySQL**: Ver 14.14

There is no option to link a slug to an existing field.

![image](https://user-images.githubusercontent.com/6013068/94472503-24a95c80-01cb-11eb-819c-23b8b3bc3bff.png)

**Comment by maartenvn** [comment]
https://github.com/directus/next/blob/282746a68fc7215bcdb3b8582860a5441c729d58/app/src/interfaces/slug/index.ts#L11

There is no option for "Mirrored Field" like in v8.
Maybe instead of a string like in v8 a dropdown could be used to select the field.
Or even better would be to use `{{ mustache }}`-syntax so you could have the option to use multiple fields:

- eg: `{{ title }}-{{ id }}`

**Comment by rijkvanzanten** [comment]
We're trying to think through how to implement this in a way that the API can enforce it as well. By adding an interface option and handling it on the client side, this would only work when using the Directus app to create an item, but will be ignored when creating an item through the API directly

**Comment by nickrum** [comment]
@rijkvanzanten have you considered implementing this combined with #353?

**Comment by rijkvanzanten** [comment]
@nickrum Yes and no. Calculated fields could be used to dynamically generate out a slug whenever you _read_ an item, but it won't help generating and saving a slug on _create_ 🤔

**Comment by nickrum** [comment]
But does it really have to be generated and saved on _create_? Wouldn't it be enough if a slug is just a dynamically generated alias? The only downside i can see is that slugs couldn't be set manually anymore which could be mitigated by additionally using a regular string field (and possibly some Interface to help with the UX). Maybe I am overlooking the obvious here 🤔

**Comment by rijkvanzanten** [comment]
@nickrum The problem with dynamically generated slugs is that they change whenever the linked field changes. With a slug-on-create, you only generate the slug once, which ensures that you don't end up with dead links in your project whenever the actual linked field changes

**Comment by nickrum** [comment]
Oh I see, that's unfortunate. Thanks for the explanation @rijkvanzanten.

**Comment by prikr** [comment]
Is there any update on this @rijkvanzanten?

**Comment by benhaynes** [comment]
[**Unfortunately, we can not provide estimates on features or fixes**](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

**Comment by prikr** [comment]
@benhaynes crystal clear!

**Comment by benhaynes** [comment]
Could this be solved with extension bundles? Then we could bundle an event hook with this interface that handles the API/logic side, right? @rijkvanzanten

**Comment by rijkvanzanten** [comment]
Yeah! That's one way :)

**Comment by jordie23** [comment]
> We're trying to think through how to implement this in a way that the API can enforce it as well. By adding an interface option and handling it on the client side, this would only work when using the Directus app to create an item, but will be ignored when creating an item through the API directly

Personally I don't think this would be a problem. Maybe I'm just thinking of my own use case, I get you guys need to think more broadly, but I'd see this ability as just a way to help make it easier to create the content item in the app. I think living in the client side app is the right place for that. If I were to create an item through an API request I'd find it annoying to not be able to manually set this value.

I guess I'm basing my concept of it on how wordpress does it. If you haven't touched the slug field, it bases the value on the title field and changes as you change the title. You can click to edit the slug field and from then on it won't auto-generate from the title field. The final value would essentially be a manually curated value, so you could expect the API requirements for the field would be to pass in a manually created value.

This reason I see it this way is that I see the slug is a fairly "technical" field, as in it's not a field that the content writers usually concern themselves with (when it's usually based on the title) so it'd be very helpful if this was a field that they rarely have to touch but can if they need to.

**Comment by benhaynes** [comment]
The current thinking is to merge this interface with the `text input` (soon to be called `input`) interface. We'll see what Rijk comes up with for the API logic side, but if that doesn't work out, then an App-only solution would suffice. Ideally Interfaces have an API  side too though.

**Reply by pwFoo** [reply]
Computed field values based on other fields would be great. Following the duscussion here.

**Comment by cdwmhcc** [comment]
This slug in the v8 version is automatically generated based on a certain field, which is very convenient at the time of creating an article, but it caused a lot of trouble to me later.

I often update the title of the article, and it will change accordingly, causing some permalinks to fail.

In v9, it is regarded as an independent field and is not associated with other fields. I think it is correct.

Of course, if there is a setting, it is only associated with a certain field when it is created, which is also a good solution.

**Reply by benhaynes** [reply]
I believe the thinking is that this would be a standard text-input, but include `on create` and `on update` options for setting the slug value based on another field. Just need to think through the most elegant way to get this functionality.

**Comment by emrahnazif** [comment]
it's clear that everybody has a different use case. 
What I believe can be a good for most cases is:

case app:
- While creating an item, slug can mirror a field (or if empty make it equal to slug of mirrored field), if manually changed, thats what it is.
- while updating an item, slug stays as is in all cases. If mirrored from field changes maybe a "refresh" icon can appear near the slug field to improve UX.

case api:
- while creating an item, if slug field is empty make it slug value of mirrored field, else value equal to slug field value. 
- while updating an item, if slug variable passed and not empty, update the field, else do nothing.

case api also corresponds to app as it should be.

So in all cases slug shouldn't be null unless mirrored field is null.

**Comment by rijkvanzanten** [comment]
The question we'll have to resole first is: how can we make the "field transforms" extendable (eg how to plug-in more than just the existing things like "current user" and secondly (and more importantly), how can we make those _configurable_. The "create a slug if it doesn't exist" is a perfect candidate for a pluggable transform type, but we currently don't have any UI / storage for it's options (in the slugs case "mirrored field").

@benhaynes I'm wondering if we should start to consider these "transforms" it's own extension type which can render options in field settings  🤔

**Reply by benhaynes** [reply]
Yes, 100%. To me, this is like _Event Hooks_, but scoped to fields.

**Reply by benhaynes** [reply]
If you create logic and save it as a file (extension), that file name (key) could be available in the `onCreate` and `onUpdate` dropdowns. Could even offer a way to upload these via the FIle Library... though that seems risky. ;)

**Reply by tylerforesthauser** [reply]
> If you create logic and save it as a file (extension), that file name (key) could be available in the `onCreate` and `onUpdate` dropdowns.

This seriously sounds amazing!

**Comment by ulvidamirli** [comment]
Are there any updates on this issue?

**Comment by danilopolani** [comment]
*(Note: this is gonna be a long comment)*

I love the idea of `onCreate` and `onUpdate` suggested by @benhaynes ! From what I understand it would not be a real-time solution tho, only server-side, right? I mean when I edit the `title` field, I'd love to see (if empty) the `slug` field updated and not to wait until I save the item.
Maybe some logic of [`interfaces`](https://docs.directus.io/extensions/interfaces/) could be reused for this? For example, let's say we call them `filters`, following the interfaces develop + build process, this could be example for an hypotetic "Slug from Title" filter.
Note that I've used a function + object instead of a Class to keep it consistent with Hooks, but this suits a Class as well.

I'll break down this code in the comments for better readability. I don't know if this code could actually work or not, it's just how I think it could be powerful. This would be a huge feature, a lot of new code and tests necessary, so don't take it too seriously haha.

<details>
  <summary>Long and boring code alert (expand to see)</summary>

  ```js
  const slugify = require('slugify');
  const axios = require('axios');
  
  interface Prop {
    name: string
    value: any
  };
  
  interface Auth {
    token: string
    user: User
  };
  
  /**
   * Verify a slug availability.
   *
   * @param  {number} itemId
   * @param  {string} originalValue
   * @param  {string} collectionName
   * @param  {string} authToken
   * @return {string}
   */
  const verifySlugAvailability = async function(itemId, fieldValue, collectionName, authToken) {
    let exists = false;
    let i = 1;
    let slug = fieldValue;
  
    while (exists) {
      const collectionCount = await axios.get(`/items/${collectionName}?&aggregate[count]=id&filter[id][_neq]=${itemId}&filter[${fieldName}][_eq]=${slug}`, {
        headers: { Authorization: `Bearer ${authToken}` }
      });
  
      if (collectionCount.data.data.id === 0) {
        exists = true;
      } else {
        slug = `${fieldValue}-${i}`;
        i++; 
      }
    }
  
    return slug;
  }
  
  export default {
  
    /**
     * Setup the filter inputs when editing the model schema fields.
     *
     * @param  {FilterManager} filterManager
     * @return {FilterManager}
     */
    setup(filterManager) {
      return filterManager.addInput({
        label: 'Mirror field',
        name: 'mirrorField',
        default: 'defaultValue',
        placeholder: 'Placeholder',
      });
    },
  
    /**
     * Handle the input changes.
     *
     * @param  {object} params0
     * @param  {string} params0.fieldName
     * @param  {any}    params0.fieldValue
     * @param  {Form}   params0.form
     * @param  {Prop[]} params0.props
     * @param  {Auth}   params0.user
     * @return {object}
     */
    onInput({ fieldName, fieldValue, form, props, user }) {
      return {
        /**
         * Watch mirror field to generate correct slug.
         */
        [props.mirrorField]: () => {
          form.setValue(fieldName, !form.getOriginal(fieldName) && !form.isDirty(fieldName) ? slugify(form.getValue(props.mirrorField)) : fieldValue),
        },
  
        /**
         * Slugify input value on change.
         */
        [fieldName]: () => {
          form.setValue(fieldName, slugify(fieldValue));
        },
      };
    },
  
    /**
     * Handle the blur of the input.
     *
     * @param  {object} params0
     * @param  {string} params0.fieldName
     * @param  {any}    params0.fieldValue
     * @param  {any}    params0.itemId
     * @param  {Form}   params0.form
     * @param  {Prop[]} params0.props
     * @param  {Auth}   params0.user
     * @return {object}
     */
    onBlur({ fieldName, fieldValue, itemId, form, props, user }) {
      return {
        /**
         * Verify slug availability on blur of mirror field.
         */
        [props.mirrorField]: async () => {
          form.setValue(fieldName, await verifySlugAvailability(itemId, fieldValue, form.getCollectionName(), auth.token));
        },
  
        /**
         * Verify slug availability on blur of the field tiself.
         */
        [fieldName]: async () => {
          form.setValue(fieldName, await verifySlugAvailability(itemId, fieldValue, form.getCollectionName(), auth.token));
        },
      };
    },
  
  };
  ```
</details>

**Reply by danilopolani** [reply]
This does three things:

1. With `setup()` it registers the incoming props from the field options when editing a model schema. The `FilterManager` class has only a few methods, like `addInput` and `getInputs()` to build the output.

<details>
  <summary>Expand to see the code</summary>

  ```js
  setup(filterManager) {
    return filterManager.addInput({
      label: 'Mirror field',
      name: 'mirrorField',
      default: 'defaultValue',
      placeholder: 'Placeholder',
    });
  },
  ```

</details>

----

2. It handles the `change`/`input` events of different fields, in order the given mirrored field from user (`props.mirrorField`) and the field to which the filter was applied (`fieldName`). In the JSDoc there are all the available arguments, some of them of course don't exist (yet?) and are just mocks, like `Auth`, `Prop` or `Form`.
2.1. When the `mirrorField` (e.g. "Title") changes, it will generate a new slug ONLY if the original field was empty and is not dirty, to not override user edits (`isDirty()`) and to not override original value (`getOriginal()`).
2.2. When the `fieldName` (e.g. "Slug") changes, just slugify it, like the "Make URL-safe" flag does.

<details>
  <summary>Expand to see the code</summary>

  ```js
  onInput({ fieldName, fieldValue, form, props, user }) {
    return {
      /**
       * Watch mirror field to generate correct slug.
       */
      [props.mirrorField]: () => {
        form.setValue(fieldName, !form.getOriginal(fieldName) ? slugify(form.getValue(props.mirrorField)) : fieldValue),
      },
  
      /**
       * Slugify input value on change.
       */
      [fieldName]: () => {
        form.setValue(fieldName, slugify(fieldValue));
      },
    };
  },
  ```

</details>

----

3. Similar to `onInput`, with `onBlur` we watch the blur event of the two inputs to check if the slug already exists and it's not the current item through the `verifySlugAvailability()` function.

<details>
  <summary>Expand to see the code</summary>

  ```js
  onBlur({ fieldName, fieldValue, itemId, form, props, user }) {
  return {
    /**
     * Verify slug availability on blur of mirror field.
     */
    [props.mirrorField]: async () => {
      form.setValue(fieldName, await verifySlugAvailability(itemId, fieldValue, form.getCollectionName(), auth.token));
    },
  
    /**
     * Verify slug availability on blur of the field tiself.
     */
    [fieldName]: async () => {
      form.setValue(fieldName, await verifySlugAvailability(itemId, fieldValue, form.getCollectionName(), auth.token));
    },
  };
  ```

</details>

**Reply by danilopolani** [reply]
Finally, the `Form` class is how to actually interact with the Vue inputs and form. I imagine something like:

```js
class Form {
  /**
   * Check if a field has been changed from the user.
   * It will return false if the field value has been changed through code.
   *
   * @param  {string} fieldName
   * @return {bool}
   */
  isDirty(fieldName) {
    return true;
  }

  /**
   * Get the original field value.
   *
   * @param  {string} fieldName
   * @return {any}
   */
  getOriginal(fieldName) {
    return 'foo';
  }

  /**
   * Get the current value of a field.
   *
   * @param  {string} fieldName
   * @return {any}
   */
  getValue(fieldName) {
    return 'foo';
  }

  /**
   * Set the value of a field.
   *
   * @param  {string} fieldName
   * @param  {any} fieldValue
   * @return {void}
   */
  setValue(fieldName, fieldValue) {
    // Do some magic here
  }
}
```

**Comment by kentkravitz** [comment]
Was this work ever finished?  It's a much needed feature!  Thanks

**Reply by joselcvarela** [reply]
@kiwimonk I don't believe so, but we are open for a PR taking into account the above things 😁

**Reply by andrenarchy** [reply]
Also see https://github.com/directus/directus/discussions/10986#discussioncomment-1977090 for how to tap into other fields.

**Comment by CanRau** [comment]
[Keystones hooks](https://keystonejs.com/docs/guides/hooks) could be an inspiration which've let me implement [my own slug logic pretty easily](https://www.canrau.com/en/cms-thoughts#keystone-js)

**Reply by rijkvanzanten** [reply]
You could do the exact same thing as a filter hook in Directus 🙂 

https://docs.directus.io/extensions/hooks/#filter

**Reply by rijkvanzanten** [reply]
I read this feature request as an app-side thing though. You'd want a slug field that updates live when somebody enters a title value into a different field.

**Reply by danilopolani** [reply]
I believe that 90% of this feature request would be used for that. Maybe reintroducing the slug field would be easier for now? I don't remember why it was removed in v9

**Reply by rijkvanzanten** [reply]
> I don't remember why it was removed in v9

It's mostly because the v8 one was often a source of confusion. The app was in charge of auto-generating the slug, and sending it up to the API, which meant that it wouldn't work if you posted to the API directly.

**Reply by danilopolani** [reply]
@rijkvanzanten would make sense to recreate that input with a flag/checkbox to automatically generate the slug for API direct calls too? I don't know if that would even be possible tho.

When the flag is checked and user didn't provide a custom slug, a new one will be automatically generated. If the flag is off, the "required" validation will fail; if the user-provided slug already exists, the "unique" validation will fail.

**Reply by rijkvanzanten** [reply]
> would make sense to recreate that input with a flag/checkbox to automatically generate the slug for API direct calls too? I don't know if that would even be possible tho.

Yeah! But that's where the technical difficulty shows up. We do currently have some field level transformations (like save current user ID, generate UUID, etc), which this could be one of (though it would need some sort of configuration). That being said, I'd rather come up with some sort of system where we can make those transformations extensions as well, so it's not just limited to a handful of these proprietary ones 🤔

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
Flows as an option/work around
- drawback, only calculates on save
- new feature --> way to preview calculated options before save ...

Should Have:
- persisted values
- calculates dynamically as fields are updated
- Add a Slugify operation with documentation on persisted vs on-demand
